gdbrsp.c

```diff
@@ -0,0 +1,77 @@
+#include <assert.h>
+#include "mini-gdbstub/include/gdbstub.h"
+#include "riscv.h"
+
+static size_t rv_read_reg(void *args, int regno)
+{
+    struct riscv_t *rv = (struct riscv_t *) args;
+
+    if (regno < 32) {
+        return rv_get_reg(rv, regno);
+    } else if (regno == 32) {
+        return rv_get_pc(rv);
+    }
+
+    return -1;
+}
+
+static void rv_read_mem(void *args, size_t addr, size_t len, void *val)
+{
+    struct riscv_t *rv = (struct riscv_t *) args;
+
+    for (size_t i = 0; i < len; i++)
+        *((uint8_t *) val + i) = rv_get_mem(rv, addr);
+}
+
+static gdb_action_t rv_cont(void *args)
+{
+    struct riscv_t *rv = (struct riscv_t *) args;
+    const uint32_t cycles_per_step = 1;
+    riscv_word_t bp_addr;
+
+    for (; !rv_has_halted(rv);) {
+        if (rv_get_breakpoint(rv, &bp_addr) && (rv_get_pc(rv) == bp_addr)) {
+            break;
+        }
+        rv_step(rv, cycles_per_step);
+    }
+
+    return ACT_RESUME;
+}
+
+static gdb_action_t rv_stepi(void *args)
+{
+    struct riscv_t *rv = (struct riscv_t *) args;
+    rv_step(rv, 1);
+    return ACT_RESUME;
+
+}
+
+static bool rv_set_bp(void *args, size_t addr, bp_type_t type)
```

Comment:
Can you make use of the APIs in `map.h` for updating breakpoints?

Reply:
I am wondering if we could support this in another patch, which will enable us to have unlimited numbers of breakpoint objects for the emulator using the `map.h`.

Reply:
In fact, I am also considering if we can implement the breakpoint interface using the trap mechanism of the emulator instead of naively comparing the program counter to the candidate breakpoints in every step, which may have bad scalability. That's also a reason that I just leave a simple approach for breakpoint here.

Reply:
Add informative `FIXME` and/or `TODO` in current source files to address further improvements and practical considerations.